example-utils/usubscription-static-file/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+/********************************************************************************
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+#![allow(clippy::mutable_key_type)]
+
+use async_trait::async_trait;
+use serde_json::Value;
+use std::collections::HashSet;
+use std::fs::{self, canonicalize};
+use std::path::PathBuf;
+use std::str::FromStr;
+use up_rust::core::usubscription::{
+    FetchSubscribersRequest, FetchSubscribersResponse, FetchSubscriptionsRequest,
+    FetchSubscriptionsResponse, NotificationsRequest, SubscriberInfo, Subscription,
+    SubscriptionRequest, SubscriptionResponse, USubscription, UnsubscribeRequest,
+};
+use up_rust::{UStatus, UUri};
+
+pub struct USubscriptionStaticFile {
+    static_file: PathBuf,
+}
+
+impl USubscriptionStaticFile {
+    pub fn new(static_file: Option<PathBuf>) -> Self {
+        // Set static_file to static_file if it has value, else set to "static-configs/testdata.json"
+        let static_file =
+            static_file.unwrap_or_else(|| PathBuf::from("static-configs/testdata.json"));
+        USubscriptionStaticFile { static_file }
+    }
+}
+
+#[async_trait]
+impl USubscription for USubscriptionStaticFile {
+    async fn subscribe(
+        &self,
+        subscription_request: SubscriptionRequest,
+    ) -> Result<SubscriptionResponse, UStatus> {
+        todo!()
+    }
+
+    async fn fetch_subscriptions(
+        &self,
+        fetch_subscriptions_request: FetchSubscriptionsRequest,
+    ) -> Result<FetchSubscriptionsResponse, UStatus> {
+        // Reads in a file and builds it into a subscription_cache data type
+        // This is a static file, so we will just return the same set of subscribers
+        // for all URIs
+        println!(
+            "fetch_subscriptions for topic: {}",
+            fetch_subscriptions_request.subscriber()
+        );
+
+        let subscription_json_file = self.static_file.clone();
+
+        let mut subscriptions_vec = Vec::new();
+
+        let empty_fetch_response = FetchSubscriptionsResponse {
+            ..Default::default()
+        };
+
+        println!("subscription_json_file: {:?}", subscription_json_file);
+
+        match canonicalize(subscription_json_file) {
```

Comment:
We can avoid the horizontal pyramid by making use of the `let Ok(foo) = ... else {};` and `let Some(foo) = ... else {};`

```rust
let Ok(subscription_json_file) = canonicalize(subscription_json_file).map_err(|e| {
  UStatus::fail_with_code(UCode::INVALID_ARGUMENT, format!("Static subscription file not found: {e:?}"))
})?;
// ... and so on for the rest ...
```

Reply:
I tried to do what you asked, take a look and lmk